File: php/testcode/massassign_020.php
<?php
require_once __DIR__ . '/shared.php';

// vuln-code-snippet start php_massassign_fill_no_fillable
function massassign020(BenchmarkRequest $req): BenchmarkResponse {
    $data = $req->param('data');
    $decoded = json_decode($data, true);
    $model = new stdClass();
    foreach ($decoded as $k => $v) {
        $model->$k = $v; // vuln-code-snippet vuln-line php_massassign_fill_no_fillable
    }
    return BenchmarkResponse::ok('filled');
}
// vuln-code-snippet end php_massassign_fill_no_fillable

File: ruby/testcode/massassign_020.rb
require_relative 'shared'

User = Struct.new(:name, :email, :role, :bio, :admin, keyword_init: true)
class << User; def create(h); new(**h.transform_keys(&:to_sym).slice(:name,:email,:role,:bio,:admin)); end; end

# vuln-code-snippet start ruby_massassign_permit_bang_nested
def handler(req)
  # permit! used upstream -- all attributes allowed through
  profile = Profile.create(req.post_data) # vuln-code-snippet vuln-line ruby_massassign_permit_bang_nested
  BenchmarkResponse.json({ ok: true })
end
# vuln-code-snippet end ruby_massassign_permit_bang_nested
